src/serde.rs

```diff
@@ -0,0 +1,128 @@
+use crate::HashBag;
+use core::fmt;
+use core::hash::{BuildHasher, Hash};
+use core::marker::PhantomData;
+use serde::de::{SeqAccess, Visitor};
+use serde::ser::{SerializeSeq, Serializer};
+use serde::Deserializer;
+use serde::{Deserialize, Serialize};
+
+pub(crate) struct HashBagVisitor<T, S> {
+    marker: PhantomData<fn() -> HashBag<T, S>>,
+}
+
+impl<T, S> HashBagVisitor<T, S>
+where
+    T: Eq + Hash,
+    S: BuildHasher + Clone,
+{
+    fn new() -> Self {
+        HashBagVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'de, T, S> Visitor<'de> for HashBagVisitor<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    type Value = HashBag<T, S>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.write_str("a HashBag")
+    }
+
+    fn visit_seq<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: SeqAccess<'de>,
+    {
+        let mut bag: HashBag<T, S> =
+            HashBag::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
+
+        while let Some(entry) = access.next_element::<(T, usize)>()? {
+            bag.insert_many(entry.0, entry.1);
```

Comment:
I wonder if this should error if the key already exists (by inspecting the return value of `insert_many` — what do you think?

Reply:
That depends. What assumptions do we want to make about the incoming data? My assumptions would be that the keys were properly `T` (de)serialized (i.e. their implementations are correct) and the user is aware of the possibility of duplicates. I think we should follow the behavior of other hashed collections. Neither `HashMap` nor `HashSet` returns an error if there are repeat keys/values. Lastly, I don't think we should provide an outcome that isn't in the `HashBag` API. There is not a, say, `swap` method that swaps the count of a key with another count.

Putting this all together, I think we should assume (with documentation) that the user ought to be aware of the potential for duplicate keys in their data, and we should treat these duplicates like the other hashed collections do, by using existing insert methods. Anything more specialized than that is up to the user (for example, take only the last occurrence of a key/count pair).

Reply:
That seems reasonable to me!